lib/spack/spack/compilers/msvc.py

```diff
@@ -292,6 +295,16 @@ def setup_custom_environment(self, pkg, env):
             else:
                 env.set_path(env_var, int_env[env_var].split(os.pathsep))
 
+        # certain versions of ifx (2021.3.0:2023.1.0) do not play well with env:TMP
+        # that has a "." character in the path
+        # Work around by pointing tmp to the stage for the duration of the build
+        if self.fc and Version(self.fc_version(self.fc)).satisfies(
+            VersionRange("2021.3.0", "2023.1.0")
+        ):
+            new_tmp = pathlib.Path(pkg.stage.path) / "tmp"
```

Comment:
This path is generally controlled by spack (one exception is `spack develop`ed packages) so there's not much chance of name clash there, but we might want a mechanism to ensure that "tmp" (or whatever we choose) doesn't exist. You could repeatedly `try/except os.mkdir` (which avoids race conditions of `if not os.path.exists...`). This might make a good function to add to `filesystem.py`:

```
def make_a_new_dir(root, prefix):
  index = 0
  while True:
    try:
      os.mkdir(prefix + index)
      break
    except FileExists
      pass

...

make_a_new_dir(pkg.stage.path, "msvc-tmp")
```

Reply:
Sounds worthwhile (although I probably rename to `safe_make_a_new_dir` to make it obvious why this is a method we bothered adding)